ccm_web/client/src/pages/GradebookCanvas.tsx

```diff
@@ -0,0 +1,390 @@
+  }, [file])
+
+  const parseUpload = (file: File|undefined): void => {
+    if (file === undefined) {
+      resetPageState()
+      return
+    }
+    // This results in an error on the 2nd "header" row for possible scores
+    parse<GradebookRecord>(file, {
+      header: true,
+      skipEmptyLines: true,
+      complete: function (results) {
+        handleParseComplete(results)
+      }
+
+    })
+  }
+
+  const resetPageState = (): void => {
+    setPageState({ state: GradebookCanvasPageState.Upload })
+  }
+
+  const getOutputFilename = (file: File | undefined): string => {
+    if (file === undefined) return ''
+    const splitName = file.name.split('.')
+    const filenameIndex = splitName.length >= 2 ? splitName.length - 2 : 0
+    splitName[filenameIndex] = splitName[filenameIndex] + '-geff'
+    return splitName.join('.')
+  }
+
+  const setCSVtoDownload = (data: GradebookRecord[]): void => {
+    let csvContent = 'data:text/csv;charset=utf-8,'
+    data.forEach(function (record, index) {
+      csvContent += record['SIS Login ID'] + ',' + record['Final Grade'] + (index < data.length ? '\n' : '')
+    })
+    setDownloadData({ data: encodeURI(csvContent), fileName: getOutputFilename(file) })
+  }
+
+  const handleNoLetterGradesError = (): void => {
+    setPageState({ state: GradebookCanvasPageState.InvalidUpload, errorMessage: [<Typography key='0'><Link href='#'>Grading Scheme in settings</Link> needs to be check marked for letter grade to appear in the CSV file.</Typography>, <Typography key='1'>{renderUploadAgainButton()}</Typography>] })
+  }
+
+  const handleRowLevelInvalidationError = (errorMessage: JSX.Element[], invalidations: GradebookRowInvalidation[]): void => {
+    setPageState({ state: GradebookCanvasPageState.InvalidUpload, invalidations: invalidations, errorMessage: errorMessage })
+  }
+
+  const handleParseSuccess = (grades: GradebookRecord[]): void => {
+    setPageState({ state: GradebookCanvasPageState.Confirm, grades: grades })
+    setCSVtoDownload(grades)
+  }
+
+  const handleParseComplete = (results: ParseResult<GradebookRecord>): void => {
+    const data = results.data.slice(1) // The first row is possible scores
+
+    if (data[0]['Final Grade'] === undefined) {
+      handleNoLetterGradesError()
+      return
+    }
+
+    let invalidations: GradebookRowInvalidation[] = []
+
+    const gradeMismatchValidator = new CurrentAndFinalGradeMatchGradebookValidator()
+
+    data.forEach(record => {
+      invalidations = invalidations.concat(gradeMismatchValidator.validate(record, data.indexOf(record) + 1 + 2)) // Add 2 because of 2 header rows
+    })
+
+    if (invalidations.length > 0) {
+      handleRowLevelInvalidationError([<div key='0'>There are blank cells in the gradebook. Please enter 0 or EX (for excused) for any blank cells in the gradebook and export a new CSV file.</div>], invalidations)
+    } else {
+      handleParseSuccess(data)
+    }
+  }
+
+  const renderUploadHeader = (): JSX.Element => {
```

Comment:
Similar case where I wonder if it should be a React Component, instead of a function.